GpsDataCapturer/wear/src/main/java/com/google/gpsdatacapturer/GpsMainActivity.java

```diff
@@ -0,0 +1,235 @@
+
+        locationManager = (LocationManager) getSystemService(Context.LOCATION_SERVICE);
+
+        requestLocationPermissionsIfNotGranted();
+        requestReadWritePermissionsIfNotGranted();
+
+        checkGpsStatus();
+        setGpsEnabled();
+
+        startAndBindGpsDataCaptureService();
+
+        final Button startAndStopButton = (Button) findViewById(R.id.start_stop_button);
+        startAndStopButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                if (isStart) {
+                    onStartButtonClick(v);
+                    startAndStopButton.setText(R.string.stop_capture);
+                    isStart = false;
+                } else {
+                    onStopButtonClick(v);
+                    startAndStopButton.setText(R.string.start_capture);
+                    isStart = true;
+                }
+            }
+        });
+
+        // Enables Always-on
+        setAmbientEnabled();
+    }
+
+    @Override
+    protected void onStart() {
+        super.onStart();
+        startAndBindGpsDataCaptureService();
+    }
+
+    @Override
+    protected void onStop() {
+        super.onStop();
+        stopAndUnbindGpsDataCaptureService();
+        isBound = false;
+    }
+
+    private void requestLocationPermissionsIfNotGranted() {
+        boolean granted = hasUserGrantedPermission(Manifest.permission.ACCESS_FINE_LOCATION)
+                && hasUserGrantedPermission(Manifest.permission.ACCESS_COARSE_LOCATION);
+        if (!granted) {
+            ActivityCompat.requestPermissions(this,
+                    new String[]{Manifest.permission.ACCESS_FINE_LOCATION,
+                            Manifest.permission.ACCESS_COARSE_LOCATION},
+                    LOCATION_REQUEST_CODE);
+        }
+
+    }
+
+    private void requestReadWritePermissionsIfNotGranted() {
+        boolean granted = hasUserGrantedPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
+                && hasUserGrantedPermission(Manifest.permission.READ_EXTERNAL_STORAGE);
+        if (!granted) {
+            ActivityCompat.requestPermissions(this,
+                    new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE,
+                            Manifest.permission.READ_EXTERNAL_STORAGE},
+                    LOCATION_REQUEST_CODE);
+        }
+    }
+
+    /**
+     * Check if the user grant permissions required to run the app
+     *
+     * @param permissionName
+     * @return return true if permission granted and false if not
+     */
+    private boolean hasUserGrantedPermission(String permissionName) {
+        boolean granted = ActivityCompat.checkSelfPermission(this, permissionName) == PackageManager.PERMISSION_GRANTED;
+        Log.d(TAG, "Permission " + permissionName + " : " + granted);
+        return granted;
+    }
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
```

Comment:
are we just logging permissions here? usually this is place where you will continue action that was delayed for user to grant permission. 

Otherwise if permission check doesn't have action context, and  whole activity is not possible without all permissions (looks like the case here) then should show toast and close/disable activity.